runtime/main/src/main/java/org/qbicc/runtime/main/VMHelpers.java

```diff
@@ -175,16 +175,206 @@ static void raiseHeapDeserializationError() {
     }
 
     // TODO: mark this with a "NoInline" annotation
-    static void raiseIncompatibleClassChangeError() { throw new IncompatibleClassChangeError(); }
+    static void raiseIncompatibleClassChangeError() {
+        throw new IncompatibleClassChangeError(); 
+    }
 
     // TODO: mark this with a "NoInline" annotation
-    static void raiseNegativeArraySizeException() { throw new NegativeArraySizeException(); }
+    static void raiseNegativeArraySizeException() {
+        throw new NegativeArraySizeException(); 
+    }
 
     // TODO: mark this with a "NoInline" annotation
     static void raiseNullPointerException() {
         throw new NullPointerException();
     }
 
     // TODO: mark this with a "NoInline" annotation
-    static void raiseUnsatisfiedLinkError() { throw new UnsatisfiedLinkError(); }
-}
+    static void raiseUnsatisfiedLinkError() {
+        throw new UnsatisfiedLinkError(); 
+    }
+ 
+    // TODO: mark with "must be build time initialized" annotation
+    static class ClinitState {
+        static final byte STATE_UNINITIALIZED = 0;
+        static final byte STATE_INPROGRESS = 1;
+        static final byte STATE_INITIALIZED = 2;
+        static final byte STATE_FAILED = 3;
+
+        Thread initializerThread;
+        Throwable errorInInitializer;
+        // Must be one of:
+        // 0 - uninitialized
+        // 1 - in-progress
+        // 2 - initialized
+        // 3 - failed
+        byte initializedState;
```

Comment:
How does this byte relate to the array `i8[]` of class stats defined above?

Reply:
The `i8[]` tracks the initialization state for buildtime initialized classes but during the initialization process, it was easier to track the state on the java heap.  Final states - initialized / failed - will be reflected back to the native data once the process completes.

This lets us use a check on-heap object while initialization is in process and then throw it away when the class is in a terminal state.